src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,138 @@
+    between the first and last frame in the clip, both included.
+
+    This isn't the same as the number of frames in a clip!
+    Example: f means a frame in the clip, x means a frame excluded from the clip
+    num_frames_per_clip = 4
+    num_indices_between_frames = 1, clip = ffff      , span = 4
+    num_indices_between_frames = 2, clip = fxfxfxf   , span = 7
+    num_indices_between_frames = 3, clip = fxxfxxfxxf, span = 10
+    """
+    return num_indices_between_frames * (num_frames_per_clip - 1) + 1
+
+
+def clips_at_random_indices(
+    decoder: SimpleVideoDecoder,
+    *,
+    num_clips: int = 1,
+    num_frames_per_clip: int = 1,
+    num_indices_between_frames: int = 1,
+    sampling_range_start: int = 0,
+    sampling_range_end: Optional[int] = None,  # interval is [start, end).
+) -> List[FrameBatch]:
+
+    _validate_params(
+        decoder=decoder,
+        num_clips=num_clips,
+        num_frames_per_clip=num_frames_per_clip,
+        num_indices_between_frames=num_indices_between_frames,
+    )
+
+    clip_span = _get_clip_span(
+        num_indices_between_frames=num_indices_between_frames,
+        num_frames_per_clip=num_frames_per_clip,
+    )
+
+    # TODO: We should probably not error.
+    if clip_span > len(decoder):
+        raise ValueError(
+            f"Clip span ({clip_span}) is larger than the number of frames ({len(decoder)})"
+        )
+
+    sampling_range_start, sampling_range_end = _validate_sampling_range(
+        sampling_range_start=sampling_range_start,
+        sampling_range_end=sampling_range_end,
+        num_frames=len(decoder),
+        clip_span=clip_span,
+    )
+
+    clip_start_indices = torch.randint(
+        low=sampling_range_start, high=sampling_range_end, size=(num_clips,)
+    )
+
+    # We want to avoid seeking backwards, so we sort the clip start indices
+    # before decoding the frames, and then re-shuffle the clips afterwards.
+    # Backward seeks may still happen if there are overlapping clips, i.e. if a
+    # clip ends after the next one starts.
+    # TODO: We should use a different strategy to avoid backward seeks:
+    # - flatten all frames indices, irrespective of their clip
+    # - sort the indices and dedup
+    # - decode all frames in index order
+    # - re-arrange the frames back into their original clips
+    clip_start_indices = torch.sort(clip_start_indices).values
+    clips = [
+        decoder.get_frames_at(
+            start=clip_start_index,
+            stop=clip_start_index + clip_span,
+            step=num_indices_between_frames,
+        )
+        for clip_start_index in clip_start_indices
+    ]
+
+    # This an ugly way to shuffle the clips using pytorch RNG *without*
+    # affecting the python builtin RNG.
+    builtin_random_state = random.getstate()
+    random.seed(torch.randint(0, 2**32, (1,)).item())
+    random.shuffle(clips)
+    random.setstate(builtin_random_state)
+
```

Comment:
Why is it necessary that we don't affect the random module's state? I assume we're trying not to affect the random values seen by models during actual training, but why do we want to do that?

Reply:
Right, I'm on board with not hardcoding a seed. I guess what I'm confused about is why we can't just call `random.shuffle(clips)` on its own. Why do we need to set a seed before shuffling? (Without that need, I assume it's okay to _consume_ values in the RNG stream.)

Reply:
We're a pytorch library, so in general we want our RNG stream to come from pytorch, not from other RNGs.
In other words we want `torch.manual_seed(123)` to affect the RNG of the sampler.

Reply:
Ohhhh, got it, so we're porting over torch's RNG state to random. Since your last response was the aha moment for me, let's say something to that effect in a comment. :)